### PR TITLE
[MGDOBR-238] Expose SLI metrics via Fleet Manager

### PR DESCRIPTION
**Pull request by robpblake**
[MGDOBR-238](https://issues.redhat.com/browse/MGDOBR-238)

This PR updates the Fleet Manager to expose the SLI metrics that have been agreed for the SLO of RHOSE. See this document for the initial draft of those SLI/SLO: https://docs.google.com/spreadsheets/d/1nZ4wNnFuH60to0h-yzOF6voRGhOnE6yd3vSRGkTJlZc/edit#gid=0

To facilitate the above, this PR introduces a number of changes:

1. The `modifiedAt` timestamp of a Bridge/Processor is now only updated when the user requests a modification. This is so we can calculate the elapsed duration of the modification operation
2. A new `deletionRequestedAt` timestamp is introduced to a Bridge/Processor entities, so that we can calculate the elapsed duration of the delete operation. This is set when the user requests that an entity be deleted via the API of RHOSE.
3. A new `MetricsService` is introduced to centralise metric reporting and calculation
4. SLI metric names are now configured via `application.properties`
5. Support for the existing metrics relating to operations on our entities is dropped, specifically the following metrics: `manager.processor.status.change` and `manager.bridge.status.change`

Update to the Grafana Dashboards to display SLOs is in the follow up task: https://issues.redhat.com/browse/MGDOBR-570

## PR owner checklist

Please make sure that your PR meets the following requirements:

- [x] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config).  
  *Run `mvn clean verify -DskipTests` so that imports are correctly set.*
- [x] Your commit messages are clear and reference the JIRA issue e.g: "[MGDOBR-1] - $clear_explanation_of_what_you_did"
- [x] The layers in the `kustomize` folder have been updated accordingly.
- [x] All new functionality is tested
- [x] Pull Request title is properly formatted: `MGDOBR-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

## PR reviewer(s) checklist

- [ ] Check if code and Github action workflows have been modified and if the modification is safe
- [ ] If the modification is safe, add the `safe to test` label

<details>
<summary>
How to trigger pipelines and use the bots:
</summary>

- <b>Run the end to end pipeline</b>  
  Annotate the pull request with the label: `safe to test`. If you want to run the pipeline again, remove and add it again.

- <b>Rebase the pull request</b>  
  Comment with: `/rebase`.
- <b>Deploy BOT</b>
  Comment with `/deploy <dev,stable> when the PR has been merged to deploy to a target environment.

</details>

### CODE REVIEW

File: manager/src/main/java/com/redhat/service/smartevents/manager/metrics/MetricsServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+            case DELETE:
+                return Duration.between(managedResource.getDeletedAt(), ZonedDateTime.now());
+            default:
+                throw new IllegalStateException(String.format("Unable to calculate operation duration for MetricsOperation '%s'", operation));
+        }
+    }
+
+    private void recordOperationDuration(ManagedResource managedResource, MetricsOperation operation) {
+        Duration operationDuration = calculateOperationDuration(managedResource, operation);
+        meterRegistry.timer(operatonDurationMetricName, buildTags(managedResource, operation)).record(operationDuration);
```
[the remark this answers is not in the document]
Comment (robpblake) [reply manager/src/main/java/com/redhat/service/smartevents/manager/metrics/MetricsServiceImpl.java:89]: @sutaakar This PR just exposes the raw metrics (SLI) required for us to be able to support our SLO. The SLO are then an interpretation of those raw metrics in terms of what we define as "good enough" performance for the user. 

Our SLO dashboard and monitoring/alerting will be built using queries on the raw data, most likely using PromQL. In pseudocode something like `if avg($bridge_provision_duration) > 5 minutes) { sendAlert("SLO is breached")}`

By exposing just the raw metric, it makes it easier for us to adjust our SLO over time without having to change the code.

Does that answer your question?

cheers,

Rob

**Review by manstis — COMMENTED** [review]
I thought I'd submitted these comments days ago.... sorry.

**Review by manstis — APPROVED** [review]
It looks like you found the additional legacy metrics I had mentioned and duly removed.

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/AbstractWorker.java [hunk 2]
```diff
@@ -114,7 +113,6 @@ protected T load(Work work) {
 
     @Transactional
     protected T persist(T managedResource) {
-        managedResource.setModifiedAt(ZonedDateTime.now());
```
[the remark this answers is not in the document]
Comment (robpblake) [reply manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/AbstractWorker.java:117]: @manstis 

The `modifiedAt` timestamp of a Bridge/Processor is now only set when the user requests a modification to the `Processor` via the API. This is so we can calculate the total elapsed duration of the modification operation. With the `worker` and the callback from the shard updating the `modifedAt` timestamp, we lose the data on when the user actually requested a modification and the duration of the operation is always incredibly short.

Does that make sense?

cheers,

Rob

File: integration-tests/src/test/resources/features/bridge.feature [hunk 3]
```diff
@@ -12,7 +12,5 @@ Feature: Bridge tests
     Then the Bridge "mybridge" is not existing within 2 minutes
     And the Ingress of Bridge "mybridge" is not available within 1 minute
 
-    And the Manager metric 'manager_bridge_status_change_total{status="PREPARING",}' count is at least 1
-    And the Manager metric 'manager_bridge_status_change_total{status="PROVISIONING",}' count is at least 1
-    And the Manager metric 'manager_bridge_status_change_total{status="READY",}' count is at least 1
-    And the Manager metric 'manager_bridge_status_change_total{status="DELETED",}' count is at least 1
+    And the Manager metric 'managed_services_api_rhose_operation_success_count_total{operation="provision",resource="bridge",}' count is at least 1
+    And the Manager metric 'managed_services_api_rhose_operation_success_count_total{operation="delete",resource="bridge",}' count is at least 1
```
Comment (sutaakar) [inline integration-tests/src/test/resources/features/bridge.feature:16]: Can you please add here at least also:
```
    And the Manager metric 'managed_services_api_rhose_operation_count_total{operation="provision",resource="bridge",}' count is at least 1
    And the Manager metric 'managed_services_api_rhose_operation_count_total{operation="delete",resource="bridge",}' count is at least 1
```
So we have at least these metrics visibly covered.

File: integration-tests/src/test/resources/features/processor.feature [hunk 4]
```diff
@@ -27,7 +27,6 @@ Feature: Processor tests
 
     Then the Processor "myProcessor" of the Bridge "mybridge" is not existing within 2 minutes
 
-    And the Manager metric 'manager_processor_status_change_total{status="PREPARING",}' count is at least 1
-    And the Manager metric 'manager_processor_status_change_total{status="PROVISIONING",}' count is at least 1
-    And the Manager metric 'manager_processor_status_change_total{status="READY",}' count is at least 1
-    And the Manager metric 'manager_processor_status_change_total{status="DELETED",}' count is at least 1
+    And the Manager metric 'managed_services_api_rhose_operation_success_count_total{operation="provision",resource="bridge",}' count is at least 1
+    And the Manager metric 'managed_services_api_rhose_operation_success_count_total{operation="provision",resource="processor",}' count is at least 1
+    And the Manager metric 'managed_services_api_rhose_operation_success_count_total{operation="delete",resource="processor",}' count is at least 1
```
Comment (sutaakar) [inline integration-tests/src/test/resources/features/processor.feature:32]: Same here for Processors.

**Review by sutaakar — APPROVED** [review]

**Review by danielezonca — COMMENTED** [review]
Just a minor comment.
Btw please promote to `dev` it as soon as merged because there are people working on Dashbuilder that are interested to do some sample with dashboards based on these metrics :)

File: manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorServiceImpl.java [hunk 5]
```diff
@@ -219,18 +219,18 @@ public Processor updateProcessorStatus(ProcessorDTO processorDTO) {
...
 
         if (processorDTO.getStatus().equals(ManagedResourceStatus.DELETED)) {
             processorDAO.deleteById(processorDTO.getId());
+            metricsService.onOperationComplete(p, MetricsOperation.DELETE);
+        } else if (processorDTO.getStatus().equals(ManagedResourceStatus.READY)) {
+            if (p.getPublishedAt() == null) {
+                p.setPublishedAt(ZonedDateTime.now());
+                metricsService.onOperationComplete(p, MetricsOperation.PROVISION);
+            } else if (p.getModifiedAt() != null) {
+                metricsService.onOperationComplete(p, MetricsOperation.MODIFY);
```
[the remark this answers is not in the document]
Comment (robpblake) [reply manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorServiceImpl.java:231]: @r00ta I'll update the code that handles the callback from the Shard to only capture the modify operation details if a `PROVISIONING` operation is on-going. That way we'll only record the update once.

cheers,

Rob

File: manager/src/main/java/com/redhat/service/smartevents/manager/metrics/MetricsOperation.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
+package com.redhat.service.smartevents.manager.metrics;
+
+import io.micrometer.core.instrument.Tag;
+
+/*
+    Constants for the logical user API operations we are report SLI data for.
+ */
+public enum MetricsOperation {
```
[the remark this answers is not in the document]
Comment (robpblake) [reply manager/src/main/java/com/redhat/service/smartevents/manager/metrics/MetricsOperation.java:8]: @danielezonca I had considered this, but it felt better to encapsulate the multiple lifecycle state changes into a single logical user metric that maps to our proposed SLO.

From an SLO perspective, right now we need SLI for the duration and correctness of the full provision, modify and delete operations. I'd suggest we start there and if we find it's useful in the future to further decompose the granularity of the metrics, we can revisit.

File: integration-tests/src/test/resources/features/processor.feature (hunk 4 above)
Comment (robpblake) [reply integration-tests/src/test/resources/features/processor.feature:32]: Sure thing.

File: integration-tests/src/test/resources/features/bridge.feature (hunk 3 above)
Comment (robpblake) [reply integration-tests/src/test/resources/features/bridge.feature:16]: Sure thing.

File: manager/src/main/java/com/redhat/service/smartevents/manager/BridgesServiceImpl.java [hunk 7]
```diff
@@ -131,7 +130,9 @@ public void deleteBridge(String id, String customerId) {
 
         // Bridge deletion and related Work creation should always be in the same transaction
         bridge.setStatus(ManagedResourceStatus.DEPROVISION);
+        bridge.setDeletedAt(ZonedDateTime.now());
```
[the remark this answers is not in the document]
Comment (robpblake) [reply manager/src/main/java/com/redhat/service/smartevents/manager/BridgesServiceImpl.java:133]: Renamed to `deletionRequestedAt` to better capture that this is a user initiated operation that will take time to complete.

**Pull request merged**